apps/prairielearn/src/lib/copy-content.ts

```diff
@@ -0,0 +1,163 @@
+import * as path from 'node:path';
+
+import { type Response } from 'express';
+import fs from 'fs-extra';
+import { v4 as uuidv4 } from 'uuid';
+import { z } from 'zod';
+
+import * as sqldb from '@prairielearn/postgres';
+import { generateSignedToken } from '@prairielearn/signed-token';
+
+import { selectCoursesWithEditAccess } from '../models/course.js';
+
+import { config } from './config.js';
+import { type Course, type CourseInstance, type Question } from './db-types.js';
+import { idsEqual } from './id.js';
+
+export interface CopyTarget {
+  id: string;
+  short_name: string | null;
+  copy_url: string;
+  __csrf_token: string;
+}
+
+const sql = sqldb.loadSqlEquiv(import.meta.url);
+
+async function getCopyTargets(res: Response, urlSuffix: string): Promise<CopyTarget[] | null> {
+  const editableCourses = await selectCoursesWithEditAccess({
+    user_id: res.locals.user.user_id,
+    is_administrator: res.locals.is_administrator,
+  });
+
+  return editableCourses
+    .filter(
+      (course) =>
+        // The example course cannot be updated in the web interface.
+        !course.example_course &&
+        // Question copying cannot be done within the same course.
+        !idsEqual(course.id, res.locals.course.id),
+    )
+    .map((course) => {
+      const copyUrl = `/pl/course/${course.id}/${urlSuffix}`;
+
+      // The copy form will POST to a different URL for each course, so
+      // we need to generate a corresponding CSRF token for each one.
+      const csrfToken = generateSignedToken(
+        {
+          url: copyUrl,
+          authn_user_id: res.locals.authn_user.user_id,
+        },
+        config.secretKey,
+      );
+
+      return {
+        id: course.id,
+        short_name: course.short_name,
+        copy_url: copyUrl,
+        __csrf_token: csrfToken,
+      };
+    });
+}
```

Comment:
This function is nearly a straight copy of the old `setQuestionCopyTargets` function, the main change is returning a typed value rather than putting stuff on `res.locals`